pkg/operator2/idp.go

```diff
@@ -292,6 +313,77 @@ func (c *authOperator) transportForCARef(ca configv1.ConfigMapNameReference, key
 	return transportFor("", caData, nil, nil)
 }
 
+func (c *authOperator) checkOIDCPasswordGrantFlow(
+	tokenURL, clientID string,
+	caRererence configv1.ConfigMapNameReference,
+	clientSecretReference configv1.SecretNameReference,
+) (bool, error) {
+	secret, err := c.secrets.Secrets("openshift-config").Get(clientSecretReference.Name, metav1.GetOptions{})
+	if err != nil {
+		return false, fmt.Errorf("couldn't get the referenced secret: %v", err)
+	}
+
+	// check whether we already attempted this not to send unneccessary login
+	// requests against the provider
+	if cachedResult, ok := oidcPasswordChecks[secret.ResourceVersion]; ok {
+		klog.V(4).Info("using cached result for OIDC password grant check")
+		return cachedResult, nil
+	}
+
+	clientSecret, ok := secret.Data["clientSecret"]
+	if !ok || len(clientSecret) == 0 {
+		return false, fmt.Errorf("the referenced secret does not contain the 'clientSecret' key")
+	}
+
+	transport, err := c.transportForCARef(caRererence, "ca")
+	if err != nil {
+		return false, fmt.Errorf("couldn't get a transport for the referenced CA: %v", err)
+	}
+
+	// prepare the grant-checking query, "openid" scope should be enough
+	query := url.Values{}
+	query.Add("client_id", clientID)
+	query.Add("client_secret", string(clientSecret))
+	query.Add("grant_type", "password")
+	query.Add("scope", "openid")
+	query.Add("username", "test")
+	query.Add("password", "test")
+	body := strings.NewReader(query.Encode())
+
+	// prepare the request, require application/json for simple parsing the response
+	req, err := http.NewRequest("POST", tokenURL, body)
+	if err != nil {
+		return false, err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	reqBytez, err := httputil.DumpRequest(req, true)
+	klog.V(1).Infof("teh bytes be: %s", reqBytez)
+
+	client := &http.Client{Transport: transport}
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	respJSON := json.NewDecoder(resp.Body)
+	respMap := map[string]interface{}{}
+	if err = respJSON.Decode(&respMap); err != nil {
+		return false, fmt.Errorf("failed to decode response from the OIDC server: %v", err)
+	}
+
+	if errVal, ok := respMap["error"]; ok {
+		oidcPasswordChecks[secret.ResourceVersion] = errVal == "invalid_grant" // wrong password, but password grants allowed
```

Comment:
So this error is the best way to discover that oidc provider supports password auth?

Reply:
I was looking for something like "yes, I would allow this, but give me correct credentials", which is what this error type seems to be used for in the tested OIDC providers.